### PR TITLE
do not promote comparing function pointers

### PR DESCRIPTION
**Pull request by RalfJung**
This *could* break existing code that relied on fn ptr comparison getting promoted to `'static` lifetime.

Fixes https://github.com/rust-lang/rust/issues/54696

### CODE REVIEW

File: src/test/run-pass/issues/issue-54696.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// run-pass
+
+fn main() {
+    // We shouldn't promote this
+    &(main as fn() == main as fn());
+    // Also check nested case
+    &(&(main as fn()) == &(main as fn()));
+}
```
Comment (Mark-Simulacrum) [inline src/test/run-pass/issues/issue-54696.rs]: How does this test verify lack of promotion? Shouldn't we instead have a compile-fail test?

Comment (arielb1) [reply src/test/run-pass/issues/issue-54696.rs]: The promotion causes crashes, because function pointer equality can't be done in constants

**Comment by arielb1** [comment]
> This *could* break existing code that relied on fn ptr comparison getting promoted to `'static` lifetime.

That code is broken anyway

**Comment by dtolnay** [comment]
@RalfJung does this also fix #54685 or is that a different issue? That one is also comparing function pointers but the discussion is focused on the LLVM side.

**Comment by RalfJung** [comment]
@dtolnay That's a different issue, but thanks for pointing me to it.

**Pull request merged**